encoding/custom/sema_codec/encode.go

```diff
@@ -0,0 +1,1101 @@
+func (e *SemaEncoder) EncodeLocationPrefix(prefix string) (err error) {
+	char := prefix[0]
+	return e.write([]byte{char})
+}
+
+var NilLocationPrefix = "\x00"
+
+// EncodeNilLocation encodes a value that indicates that no location is specified
+func (e *SemaEncoder) EncodeNilLocation() (err error) {
+	return e.EncodeLocationPrefix(NilLocationPrefix)
+}
+
+func (e *SemaEncoder) EncodeAddressLocation(t common.AddressLocation) (err error) {
+	err = e.EncodeLocationPrefix(common.AddressLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	err = e.EncodeAddress(t.Address)
+	if err != nil {
+		return
+	}
+
+	return e.EncodeString(t.Name)
+}
+
+func (e *SemaEncoder) EncodeIdentifierLocation(t common.IdentifierLocation) (err error) {
+	err = e.EncodeLocationPrefix(common.IdentifierLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	return e.EncodeString(string(t))
+}
+
+func (e *SemaEncoder) EncodeScriptLocation(t common.ScriptLocation) (err error) {
+	err = e.EncodeLocationPrefix(common.ScriptLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	return e.write(t[:])
+}
+
+func (e *SemaEncoder) EncodeStringLocation(t common.StringLocation) (err error) {
+	err = e.EncodeLocationPrefix(common.StringLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	return e.EncodeString(string(t))
+}
+
+func (e *SemaEncoder) EncodeTransactionLocation(t common.TransactionLocation) (err error) {
+	err = e.EncodeLocationPrefix(common.TransactionLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	return e.write(t[:])
+}
+
+func (e *SemaEncoder) EncodeREPLLocation() (err error) {
+	return e.EncodeLocationPrefix(common.REPLLocationPrefix)
+}
+
+// EncodeString encodes a string as a byte array.
+func (e *SemaEncoder) EncodeString(s string) (err error) {
+	return e.EncodeBytes([]byte(s))
+}
+
+// EncodeBytes encodes a byte array.
+func (e *SemaEncoder) EncodeBytes(bytes []byte) (err error) {
+	err = e.EncodeLength(len(bytes))
+	if err != nil {
+		return
+	}
+
+	return e.write(bytes)
+}
+
+// TODO encode length with variable-sized encoding?
+//      e.g. first byte starting with `0` is the last byte in the length
+//      will usually save 3 bytes. the question is if it saves or costs encode and/or decode time
+
+// EncodeLength encodes a non-negative length as a uint32.
+// It uses 4 bytes.
+func (e *SemaEncoder) EncodeLength(length int) (err error) {
+	if length < 0 { // TODO is this safety check useful?
+		return fmt.Errorf("cannot encode length below zero: %d", length)
+	}
+
+	l := uint32(length)
+
+	return binary.Write(&e.w, binary.BigEndian, l)
+}
+
+func (e *SemaEncoder) EncodeAddress(address common.Address) (err error) {
+	return e.write(address[:])
+}
+
+func (e *SemaEncoder) write(b []byte) (err error) {
+	_, err = e.w.Write(b)
+	return
+}
+
+func EncodeArray[T any](e *SemaEncoder, arr []T, encodeFn func(T) error) (err error) {
+	// TODO save a bit in the array length for nil check?
+	err = e.EncodeBool(arr == nil)
```

Comment:
Maybe flip this? When reading the test cases above, it was surprising that `false` meant there is an array to follow. 

Maybe also differentiate between optional and non-optional arrays? This function not only encodes an array, but also handles the case where the array is nil. That means that in cases where there always is an array, an additional byte is wasted. The simplest solution might be to just always write just the size and write 0 for the nil case?

Reply:
I originally translated nil array into array of size zero but then the encoded and decoded types differed when an array was nil. I don't know if that matters outside of making tests a little more complex.

Also, I don't know if most nil-able values will ever actually be nil.